contrib/opencensus-context/opencensus/common/runtime_context/__init__.py

```diff
@@ -0,0 +1,135 @@
+
+    def __getattr__(self, name):
+        if name not in self._slots:
+            raise AttributeError('{} is not a registered context slot'
+                                 .format(name))
+        slot = self._slots[name]
+        return slot.get()
+
+    def __setattr__(self, name, value):
+        if name not in self._slots:
+            raise AttributeError('{} is not a registered context slot'
+                                 .format(name))
+        slot = self._slots[name]
+        slot.set(value)
+
+
+class _ThreadLocalRuntimeContext(_RuntimeContext):
+    _lock = threading.Lock()
+    _slots = {}
+
+    class Slot(object):
+        _thread_local = threading.local()
+
+        def __init__(self, name, default):
+            self.name = name
+            self.default = default if callable(default) else (lambda: default)
+
+        def clear(self):
+            setattr(self._thread_local, self.name, self.default())
+
+        def get(self):
+            try:
+                return getattr(self._thread_local, self.name)
+            except AttributeError:
+                value = self.default()
+                self.set(value)
+                return value
+
+        def set(self, value):
+            setattr(self._thread_local, self.name, value)
+
+    @classmethod
+    def clear(cls):
+        with cls._lock:
+            for name in cls._slots:
+                slot = cls._slots[name]
+                slot.clear()
+
+    @classmethod
+    def register_slot(cls, name, default=None):
+        with cls._lock:
+            if name in cls._slots:
+                raise ValueError('slot {} already registered'.format(name))
+            cls._slots[name] = cls.Slot(name, default)
+
+
+class _AsyncRuntimeContext(_RuntimeContext):
+    _lock = threading.Lock()
+    _slots = {}
+
+    class Slot(object):
+        def __init__(self, name, default):
+            import contextvars
+            self.name = name
+            self.contextvar = contextvars.ContextVar(name)
+            self.default = default if callable(default) else (lambda: default)
+
+        def clear(self):
+            self.contextvar.set(self.default())
+
+        def get(self):
+            try:
+                return self.contextvar.get()
+            except LookupError:
+                value = self.default()
+                self.set(value)
+                return value
+
+        def set(self, value):
+            self.contextvar.set(value)
+
+    @classmethod
+    def clear(cls):
+        with cls._lock:
+            for name in cls._slots:
+                slot = cls._slots[name]
+                slot.clear()
+
+    @classmethod
+    def register_slot(cls, name, default=None):
+        with cls._lock:
+            if name in cls._slots:
+                raise ValueError('slot {} already registered'.format(name))
+            cls._slots[name] = cls.Slot(name, default)
+
+
+RuntimeContext = _ThreadLocalRuntimeContext()
+
+if sys.version_info >= (3, 7):
```

Comment:
@bogdandrutu FYI - this is where we switch from ThreadLocal to Async (contextvard) based on the runtime version.